compiler/rustc_borrowck/src/type_check/mod.rs

```diff
@@ -2021,7 +2024,13 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
                         // and hence may contain unnormalized results.
                         let fn_sig = self.normalize(fn_sig, location);
 
-                        let ty_fn_ptr_from = tcx.safe_to_unsafe_fn_ty(fn_sig);
+                        let ty_fn_ptr_from = match ptr_cast {
+                            PointerCast::UnsafeFnPointer => tcx.safe_to_unsafe_fn_ty(fn_sig),
+                            PointerCast::DeprecatedSafeFnPointer => {
+                                tcx.unsafe_to_safe_fn_ty(fn_sig)
+                            }
+                            _ => unreachable!(),
+                        };
```

Comment:
why is that needed?

Reply:
i have:
```
CastKind::Pointer(
    ptr_cast @ (PointerCast::UnsafeFnPointer
    | PointerCast::DeprecatedSafeFnPointer),
) => {
```
followed by:
let ty_fn_ptr_from = match ptr_cast {
    PointerCast::UnsafeFnPointer => ...,
    PointerCast::DeprecatedSafeFnPointer => ...,
    _ => unreachable!(),
};
```

being a rust noob, i'm kinda hoping there's a better way here 😊 i want to match on the two variants that i've already guaranteed are the only two i can see

Reply:
ah, the branches are actually different :sweat_smile: `unsafe -> safe` vs `safe -> unsafe` xx

sry

Reply:
i'll run into the same thing for determining the message when i merge [these two arms](https://github.com/rust-lang/rust/blob/5c1f53ea37387c615c573c7a9a01822091edc41e/compiler/rustc_const_eval/src/interpret/cast.rs#L66-L86) like you suggested as well

Reply:
no worries 😊 do you know if the ```unreachable!()``` is possible to remove though? it bothers me 🤓

Reply:
i don't think it can easily be removed :sweat_smile: you could merge `PointerCast::DeprecatedSafeFnPointer` and `PointerCast::UnsafeFnPointer` into
`PointerCast::FnPointer(FnCastKind)` but I don't think that actually improves anything :sweat_smile:

Reply:
thanks! guess i'll have to live with it 😊 i'll still go ahead and merge the arms in the other two places though

Reply:
the unreachable!() is commented now